worker/raft/worker_test.go

```diff
@@ -329,7 +337,9 @@ func (s *WorkerSuite) TestNoLeaderTimeout(c *gc.C) {
 	// waits when we advance:
 	// * the loop timeout wait from starting the worker
 	// * the no leader timeout check in loop.
-	c.Assert(s.clock.WaitAdvance(10*time.Second, coretesting.LongWait, 2), jc.ErrorIsNil)
+	// The NoLeaderTimeout waits for 10s, so we increment the clock just slightly
+	// longer than that.
+	c.Assert(s.clock.WaitAdvance(10001*time.Millisecond, coretesting.LongWait, 2), jc.ErrorIsNil)
```

Comment:
Given all the other cases in the code where we advance exactly the right amount, I don't think this is the issue.

Reply:
I can see your point (essentially this changes the test from >= to >), but I think it's worth seeing if it does actually change the behaviour given that we can't reproduce it locally.

Reply:
So this was a red herring. Which we followed up with *another* red herring. (Namely, inside the code we use raft.LastContact() which uses a real time.Now() call.) That *could* be the failure, but it would only trigger if the Clock time and the time.Now() time differ by 6 seconds (no leader timeout is 4s, clock.WaitAdvance is 10s.)

However, it turned out to really be the failed "am I still the leader" check.

If you put a 1ms sleep for the node to step down after getting the Demote request. Then the request succeeds (f3.Error() is nil). But you can get to the call WaitAdvance(10s) *before* the node actually steps down as the leader. Which means that when the Worker loop checks "r.State() == raft.Leader" it gets True. And if *I* am the leader, than clearly there we aren't  missing a leader, and we shouldn't trigger NoLeaderTimeout.

So the fix really should be the "fix the loop to wait until I'm not the leader", and I've reverted the change around 10s vs 10.001s.

Reply:
Curses, Tim was right!